cs/ccxt/base/Throttler.cs

```diff
@@ -67,22 +82,89 @@ private async Task loop()
                 var elapsed = current - lastTimestamp;
                 lastTimestamp = current;
                 var tokens = (double)this.config["tokens"] + ((double)this.config["refillRate"] * elapsed);
-                this.config["tokens"] = Math.Min(tokens, (int)this.config["capacity"]);
+                this.config["tokens"] = Math.Min(tokens, (double)this.config["capacity"]);
             }
         }
 
     }
 
-    public async Task<Task> throttle(object cost2)
+    private async Task rollingWindowLoop()
+    {
+        while (this.running)
+        {
+            lock (queueLock)
+            {
+                if (this.queue.Count == 0)
+                {
+                    this.running = false;
+                    continue;
+                }
+            }
+            (Task, double) first;
+            lock (queueLock)
+            {
+                first = this.queue.Peek();
+            }
+            var task = first.Item1;
+            var cost = first.Item2;
+            var now = milliseconds();
+            var windowSize = Convert.ToDouble(this.config["windowSize"]);
+            timestamps = timestamps.Where(t => now - t.timestamp < windowSize).ToList();
```

Comment:
```suggestion
            timestamps.RemoveAll(t => now - t.timestamp >= windowSize);
```
what about this @caoilainnl ?